---
title: "ABurillo_metagenomics"
author: "ALBA BURILLO"
date: "15/3/2021"
output:
  pdf_document: default
  html_document: default
---

### METAGENOMICS FINAL TASK

In this task we are going to perform a metagenomics analysis and we are going to study the role of oral microbiome in HIV infection and clinical evolution. The dataset that we are going to used was obtained from Illumina/MiSeq sequencing of oral samples from HIV infected subjects with different degrees of periodontal disease (PD). We have metagenomic data from 250 samples including healthy and HIV-infected patients.

We will analyze an OTU table (OTUTable_Evaluation.shared) and the associated metadata (Metadata_Evaluation.txt).

```{r setup, include=FALSE}
library(dada2); packageVersion("dada2") ## Dada2 v.1.18.0
library(ShortRead); packageVersion("ShortRead") ## ShortRead 1.48.0
library(phyloseq); packageVersion("phyloseq") ## phyloseq 1.34.0
library(gridExtra)
require(digest)

suppressPackageStartupMessages(library(phyloseq))### Bioconductors
suppressPackageStartupMessages(library(vegan))### Bioconductors
suppressPackageStartupMessages(library(ggplot2))### CRAN
suppressPackageStartupMessages(library(DESeq2)) ### Bioconductors
suppressPackageStartupMessages(require(RColorBrewer))### CRAN
#######   gdata packages allows for microsofft office filetype read-in
suppressPackageStartupMessages(library(gdata))### CRAN
packageVersion("phyloseq")
packageVersion("vegan")
packageVersion("ggplot2")
packageVersion("DESeq2")
packageVersion("RColorBrewer")

#install.packages('doBy', dependencies = T )
library(doBy)


```

### LOAD THE DATA

```{r, echo= TRUE, fig.align='center'}

#### Set Working directory to the root directory where you have your data code
code_path<-("C:/Users/Albaburillo/Desktop/FINAL_TASK_METAGENOMICS/EvaluationMaterials")
code_path
setwd(code_path)

#### Read metadata
metadata<-read.table(paste0(code_path,"/Metadata_Evaluation.txt"), header = TRUE)
dim(metadata)
metadata$HIVStatus<-as.factor(metadata$HIVStatus)
metadata$PSOE <- as.factor(metadata$PSOE)
metadata$Site<- as.factor(metadata$Site)
summary(metadata)

rownames(metadata)<-metadata$SampleID

### taxonomically classify down to the genus level,
AvailableRanks<-c("Kingdom","Phylum","Class","Order","Family","Genus")

### Define the minimum number of OTU counts for a sample to be further analyzed
minSampleCountsB<-1000


data<-import_mothur(mothur_constaxonomy_file="Taxonomy_evaluation.txt",
                  mothur_shared_file="OTUTable_Evaluation.shared",
                  mothur_tree_file="phyloTree_Evaluation.tree",
                  cutoff=0.03)
### Attach the metadata
sample_data(data)<-metadata
data
### Define available ranks in taxonomy file
colnames(tax_table(data))<-AvailableRanks

par(mfrow=c(1,2))  
barplot(colSums(otu_table(data)),las=2,cex.names=0.6,main="#Counts/Sample")
data<-prune_samples(sample_sums(data)>minSampleCountsB,data)
data

### Let's plot the number of counts for the remaining samples
barplot(colSums(otu_table(data)),las=2,cex.names=0.6,main="#Counts/Sample")

### We can se that we filtered only 1 
### Now we read the phylogenetic tree into a phylogenetic object (ape package)
phylotree<-read_tree("phyloTree_Evaluation.tree")



```

### ALPHA DIVERSITY ANALYSIS

```{r, echo= TRUE, fig.align='center'}
#### Start Alpha - Diversity Analysis
### We will characterize alpha-diversity indices using a rarefied subset of 5000 counts in order to balance sampling between samples. First we try to apply a prevalence filter
x.2.0 <- data
wh0=genefilter_sample(x.2.0,filterfun_sample(function(x) x>1), A=0.01*nsamples(x.2.0))
x.2.0<-prune_taxa(wh0,x.2.0)

### We can draw a rarefaction curve using otu_table from x.2.0 5000 subset object
 
rarecurve(data.frame(t(otu_table(x.2.0))),step=20,col=sample_data(x.2.0)$HIVStatus,label=F)
rarecurve(data.frame(t(otu_table(x.2.0))),step=20,col=sample_data(x.2.0)$Site,label=F)

### Let's see what would happen if we had not applied a simple filter
x.2.0 <- data
rarecurve(data.frame(t(otu_table(x.2.0))),step=20,col=sample_data(x.2.0)$Site,label=F)

# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(x.2.0),
               MARGIN = ifelse(taxa_are_rows(x.2.0), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(x.2.0),
                    tax_table(x.2.0))
plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})
prevdf1 = subset(prevdf, Phylum %in% get_taxa_unique(x.2.0, "Phylum"))
ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(x.2.0),color=Phylum)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) + geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none")
#ggsave("Abundance_vs_Prevalence_Phylum.pdf")


```

### DIVERSITY PLOTS

```{r, echo= TRUE, fig.align='center',  fig.cap= 'The Shannon, Simpson and InvSimpson indexes of the samples regarding Site group'}

x.2.0 <- data
#We randomly subset 5000 counts from each sample and discard all samples with fewer counts
x.2.0<-rarefy_even_depth(x.2.0,3000) ##### RRarefy to 5000 counts
##########Generate Diversity Plots 

p<-plot_richness(x.2.0,"Site",measures=c("Shannon","Simpson","InvSimpson")) 
p+geom_boxplot(aes(fill=Site)) 


```

### RICHNESS PLOTS

```{r, echo= TRUE, fig.align='center', fig.cap= 'The Observed, Chao1 and ACE indexes of the samples regarding Site group'}

########Generate Richness Plots

p<-plot_richness(x.2.0,"Site",measures=c("Observed","Chao1","ACE")) 
p+geom_boxplot(aes(fill=Site)) 

```

```{r, echo= TRUE}


my.sampleData<-data.frame(sample_data(x.2.0))
er.x.2.0<-estimate_richness(x.2.0)
rownames(er.x.2.0)<-rownames(my.sampleData)

er.x.2.0<-merge(my.sampleData,er.x.2.0,by="row.names",all.x=T)
rownames(er.x.2.0)<-rownames(my.sampleData)
sample_data(x.2.0)<-er.x.2.0


### Plots are nice and look good (underline some trends, but we need statistics)
### Generate Diversity Stats and capture them in a txt file
capture.output(file="DivAndRichnessStats.txt",paste())
for(covar in c("HIVStatus","Site")){
  for (measure in c("Observed","Chao1","ACE","Shannon","Simpson","InvSimpson")){
    capture.output(file="DivAndRichnessStats.txt",paste("############",measure,"by",covar),append=T)
    if(length(unique(er.x.2.0[,covar]))==2){
      myttest<-t.test(er.x.2.0[,measure]~er.x.2.0[,covar])
      capture.output(file="DivAndRichnessStats.txt",myttest,append=T)
    }else{
      myanova<-aov(er.x.2.0[,measure]~er.x.2.0[,covar])
      capture.output(file="DivAndRichnessStats.txt",myanova,append=T)
      capture.output(file="DivAndRichnessStats.txt",summary(myanova),append=T)
      capture.output(file="DivAndRichnessStats.txt",TukeyHSD(myanova),append=T)
    }
  }
}

DivAndRichness <- read.delim("DivAndRichnessStats.txt", row.names=NULL)
SiteDivAndRichness <- DivAndRichness[61:186,1]
SiteDivAndRichness

```

### 1. Which Sampling site shows the highest Diversity? And Richness?

We can see in the diversity plots that the sampling site with the highest diversity is the Teeth. The alpha diversity measure in the three methods (Shannon, Simpson and InvSimpson) is clearly higher in the teeth, with a great difference compared to the other two sites (saliva and cheek). In the SiteDivAndRichness object we have the result of the t-tests and the anova of diversity. In the three measures (shannon, simpson and InvSimpson) we can see that the diversity at the sampling sites is significantly different,being higher in Teeth.

We can see in the richness plots that the sampling site with the highest richness is the Cheek. The alpha diversity measure in the different methods (Observed, Chao1 and ACE) is almost the same in the three sites (Saliva, Teeth and Cheek), but a little higher in the cheek. In the SiteDivAndRichness object we have the result of the t-tests and the anova of diversity. In the three measures (Observed, Chao1 and ACE) we can see that the richness at the sampling sites is not significantly different.

### TAXONOMICAL ANALYSIS

```{r, echo= TRUE, fig.align='center'}
################# Taxonomical analysis ########################
### Let's make an initial taxonomy description of the data
### Cumulative Stacked Barplots at Phylum, Genus and Species
### 3.0
x.3.0<-data
### We apply a more stringent filter, basically we are not interested in rare OTUs but on
### general main trends of taxonomical composition
wh0=genefilter_sample(x.3.0,filterfun_sample(function(x) x>5), A=0.1*nsamples(x.3.0))
x.3.0<-prune_taxa(wh0,x.3.0)
### tax_glom function agglomerates/collapses all OTU belonging to the same taxonomical level
x.3.0.phylum<-tax_glom(x.3.0,taxrank="Phylum")
x.3.0.genus<-tax_glom(x.3.0,taxrank="Genus")
x.3.0.genus<-subset_taxa(x.3.0.genus,Genus != "unclassified")
x.3.0.genus<-subset_taxa(x.3.0.genus,Genus != "Incertae_Sedis")
### Convert to data.frame for easier manipulation
ps.melt.x.3.0.phylum<-psmelt(x.3.0.phylum)
ps.melt.x.3.0.genus<-psmelt(x.3.0.genus)

### We will calculate relative abundances as proportions for further analysis
ps.melt.x.3.0.phylum$AbundanceProportion <- ave(ps.melt.x.3.0.phylum$Abundance,list(ps.melt.x.3.0.phylum[,"SampleID"]), FUN=function(L) L/sum(L))
ps.melt.x.3.0.genus$AbundanceProportion <- ave(ps.melt.x.3.0.genus$Abundance,list(ps.melt.x.3.0.genus[,"SampleID"]), FUN=function(L) L/sum(L))

reverse.levels <- function(x) {
  if(is.factor(x)) {
    x <- factor(as.character(x), levels=rev(levels(x)), ordered=TRUE)
  } else if(is.data.frame(x)) {
    for(i in seq_along(x)) {
      if(is.factor(x[,i])) {
        x[,i] <- factor(as.character(x[,i]), levels=rev(levels(x[,i])), ordered=TRUE)
      } else {
        warning(paste0('Column ', i, ' is not a factor.'))
      }
    }
  } else {
    stop(paste0('Unsupported format: ', class(x)))
  }
  return(x)
}

```

### BARPLOTS AT PHYLUM

```{r, echo= TRUE, fig.align='center'}
### Barplots at phylum. We use decreasing phylum abundances as X order
### and the phylum abundance of first sample as y order
### This careful use of X/Y order will already reveal patterns
### Define Level Order for X axis (SampleID)

my.levels=orderBy(~-AbundanceProportion,data=ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$Phylum=="Bacteroidetes",])$SampleID
ps.melt.x.3.0.phylum$SampleID<-factor(ps.melt.x.3.0.phylum$SampleID,
                                      levels=my.levels,ordered=T)

### Define Level Order for Y axis (phylum,genus,species)
my.levels<-orderBy(~-AbundanceProportion,data=ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$SampleID == as.character(my.levels[1]),])$Phylum
ps.melt.x.3.0.phylum$Phylum<-factor(ps.melt.x.3.0.phylum$Phylum,levels=my.levels,ordered=T)

### We also take care of colors
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7","#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colorCount=length(unique(ps.melt.x.3.0.phylum$Phylum))
getPalette=colorRampPalette(brewer.pal(12,"Set3"))
p<-ggplot(ps.melt.x.3.0.phylum,aes(x=SampleID,y=AbundanceProportion,order=Phylum))
p+geom_bar(stat="identity",aes(fill=Phylum))+ theme(axis.text.x = element_text(size=rel(0.5),angle = 90, hjust = 1))+labs(title="Phylum Level")+
  scale_fill_manual(values=getPalette(colorCount))+guides(fill=guide_legend(ncol=2))


```

### BARPLOTS AT GENUS

```{r, echo= TRUE, fig.align='center'}
#Barplots at genus
# Define Level Order for X axis (SampleID)
my.levels=orderBy(~-AbundanceProportion,data=ps.melt.x.3.0.genus[ps.melt.x.3.0.genus$Genus=="Veillonella",])$SampleID
ps.melt.x.3.0.genus$SampleID<-factor(ps.melt.x.3.0.genus$SampleID,
                                     levels=my.levels,ordered=T)
#Define Level Order for Y axis (genus,genus,species)
my.levels<-orderBy(~-AbundanceProportion,data=ps.melt.x.3.0.genus[ps.melt.x.3.0.genus$SampleID == as.character(my.levels[1]),])$Genus
ps.melt.x.3.0.genus$Genus<-factor(ps.melt.x.3.0.genus$Genus,levels=my.levels,ordered=T)

ps.melt.x.3.0.genus<-subset(ps.melt.x.3.0.genus,ps.melt.x.3.0.genus$AbundanceProportion>0.02)
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7","#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
colorCount=length(unique(ps.melt.x.3.0.genus$Genus))
getPalette=colorRampPalette(brewer.pal(12,"Set3"))
p<-ggplot(ps.melt.x.3.0.genus,aes(x=SampleID,y=AbundanceProportion,order=Genus))
p+geom_bar(stat="identity",aes(fill=Genus))+ theme(axis.text.x = element_text(size=rel(0.5),angle = 90, hjust = 1))+labs(title="genus Level")+
  scale_fill_manual(values=getPalette(colorCount))+guides(fill=guide_legend(ncol=2))

```

```{r, echo= TRUE, fig.align='center'}
### We can assess Differential Abundance, related to any variable in our metadata
### 3.0
x.3.0<-data
wh0=genefilter_sample(x.3.0,filterfun_sample(function(x) x>1), A=0.1*nsamples(x.3.0))
x.3.0<-prune_taxa(wh0,x.3.0)
### tax_glom function agglomerates/collapses all OTU belonging to the same taxonomical level
x.3.0.phylum<-tax_glom(x.3.0,taxrank="Phylum")
x.3.0.genus<-tax_glom(x.3.0,taxrank="Genus")

ps.melt.x.3.0.phylum<-psmelt(x.3.0.phylum)
ps.melt.x.3.0.genus<-psmelt(x.3.0.genus)

### Let's transform our data in different ways
ps.melt.x.3.0.phylum$Abundance<-floor(ps.melt.x.3.0.phylum$Abundance)
### Add a pseudocount to allow for log transformation
ps.melt.x.3.0.phylum$Abundance<-ps.melt.x.3.0.phylum$Abundance+1
ps.melt.x.3.0.phylum$AbundanceProportion <- ave(ps.melt.x.3.0.phylum$Abundance,list(ps.melt.x.3.0.phylum[,"SampleID"]), FUN=function(L) L/sum(L))
ps.melt.x.3.0.phylum$AbundanceProportionLog <- log10(ps.melt.x.3.0.phylum$AbundanceProportion)
ps.melt.x.3.0.phylum$AbundanceNorm<-floor(10000*ps.melt.x.3.0.phylum$AbundanceProportion)

ggplot(ps.melt.x.3.0.phylum,aes(x=Phylum,y=AbundanceProportionLog,fill=Phylum))+geom_boxplot()+facet_wrap(~Site)+ theme(axis.text.x = element_text(angle = 90))

ggplot(ps.melt.x.3.0.phylum,aes(x=Phylum,y=AbundanceProportionLog,fill=HIVStatus))+geom_boxplot()+facet_wrap(~Site)+ theme(axis.text.x = element_text(angle = 90))


ps.melt.x.3.0.genus$Abundance<-floor(ps.melt.x.3.0.genus$Abundance)
ps.melt.x.3.0.genus$Abundance<-ps.melt.x.3.0.genus$Abundance+1
ps.melt.x.3.0.genus$AbundanceProportion <- ave(ps.melt.x.3.0.genus$Abundance,list(ps.melt.x.3.0.genus[,"SampleID"]), FUN=function(L) L/sum(L))
ps.melt.x.3.0.genus$AbundanceProportionLog <- log10(ps.melt.x.3.0.genus$AbundanceProportion)
ps.melt.x.3.0.genus$AbundanceNorm<-floor(10000*ps.melt.x.3.0.genus$AbundanceProportion)

### To plot at Genus level we first need to select the most abundant genus
### plotting all of them is a terrible mess
meanByGenus<-aggregate(ps.melt.x.3.0.genus$AbundanceProportion,list(ps.melt.x.3.0.genus$Genus),mean)
colnames(meanByGenus)<-c("Genus","meanAbundance")
meanByGenus<-meanByGenus[with(meanByGenus, order(-meanAbundance, Genus)), ]

##### phylum
meanByPhylum<-aggregate(ps.melt.x.3.0.phylum$AbundanceProportion,list(ps.melt.x.3.0.phylum$Phylum),mean)
colnames(meanByPhylum)<-c("Phylum","meanAbundance")
meanByPhylum<-meanByPhylum[with(meanByPhylum, order(-meanAbundance, Phylum)), ]
```

### RELATIVE LOG PROPORTIONS OF MOST 10 ABUNDANT PHYLUM AND GENUS REGARDING SAMPLING SITE

```{r, echo= TRUE, fig.align='center'}
par(mfrow=c(1,2))
#### Plot relative logProportions of most 10 abundant genus
ggplot(ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$Phylum %in% meanByPhylum[1:10,1],],aes(x=Phylum,y=AbundanceProportionLog,fill=Phylum))+geom_boxplot()+facet_wrap(~Site)+ theme(axis.text.x = element_text(angle = 90))

ggplot(ps.melt.x.3.0.genus[ps.melt.x.3.0.genus$Genus %in% meanByGenus[1:10,1],],aes(x=Genus,y=AbundanceProportionLog,fill=Genus))+geom_boxplot()+facet_wrap(~Site)+ theme(axis.text.x = element_text(angle = 90))

```

### 2.Which are the bacterial genus that are enriched in teeth samples compared to both cheek and saliva?

In the boxplots of the relative logarithmic proportions of the 10 most abundant genera, we can observed that the most abundant genus in teeth samples compared to Saliva and cheek are Corynebacterium and Selenomonas.But we must perform statistics to confirm it,

```{r, echo= TRUE, fig.align='center'}
############ Start Genus comparison barplots #########################
### We have seen that there are differences but we need to see which these differences are
###Barplots for Significant Genus difference
x.4.0<-data
wh0=genefilter_sample(x.4.0,filterfun_sample(function(x) x>1), A=0.01*nsamples(x.4.0))
x.4.0<-prune_taxa(wh0,x.4.0)
x.4.0 = transform_sample_counts(x.4.0, function(x) ((x/sum(x))))
x.4.0<-tax_glom(x.4.0,taxrank="Genus")
psmelt.x.4.0.genus<-psmelt(x.4.0)

##### Barplots for significant genus related to HIVstatus
mysignificantGenus<-vector()
my.p.values.vector<-vector()
# Find significant genus (p<0.01)
mainDir<-"./"
subDir<-"StatisticalTests"
if (file.exists(subDir)){
  #setwd(file.path(mainDir, subDir))
} else {
  dir.create(file.path(mainDir, subDir))
  #setwd(file.path(mainDir, subDir))
}

### For every Genus we will calculate a Wilcoxon Rank-based test between dicotomic variable
for(genus in as.vector(unique(psmelt.x.4.0.genus$Genus))){
  mytest<-wilcox.test(Abundance~HIVStatus,data=psmelt.x.4.0.genus[psmelt.x.4.0.genus$Genus==genus,])
  capture.output(mytest,file=paste(mainDir,"/",subDir,"/",genus,"_HIVStatus_Wilcoxon.txt",sep=""))
  my.p.values.vector<-c(my.p.values.vector,mytest$p.value)
}

### We need to correct for multiple test error, we use benjamin-hochberg
my.p.values.vector.adj<-p.adjust(my.p.values.vector,method="BH")
mysignificantGenus<-unique(psmelt.x.4.0.genus$Genus)[my.p.values.vector.adj<0.05]
ps.x.4.0.genus<-psmelt.x.4.0.genus[psmelt.x.4.0.genus$Genus %in% mysignificantGenus,]

p<-ggplot(ps.x.4.0.genus,aes(x=Genus,y=Abundance))


#Plot B&W boxplot
 
p+geom_boxplot(aes(fill=Site,stat="identity"),notch=F,position="dodge")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+ scale_fill_grey(start = 0.6, end = 1)+
  theme_bw() + theme(legend.title=element_blank(),legend.text=element_text(size=22),legend.position=c(.5, .5))+
  theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black",size=0.9),
        axis.text.x = element_text(angle = 45, hjust = 1,size=8),axis.text.y = element_text(angle = 45, hjust = 1,size=8),axis.title=element_text(size=8,face="bold"))

#Plot Color barplot
p+geom_boxplot(aes(fill=Site,stat="identity"),notch=F,position="dodge")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  theme_bw() + theme(legend.title=element_blank(),legend.text=element_text(size=22),legend.position=c(.5, .5))+
  theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black",size=0.9),
        axis.text.x = element_text(angle = 45, hjust = 1,size=8),axis.text.y = element_text(angle = 45, hjust = 1,size=8),axis.title=element_text(size=8,face="bold"))+scale_y_continuous(trans='sqrt')
```

```{r, echo= TRUE, fig.align='center'}
require("ggpubr")
p+geom_boxplot(aes(fill=Site),notch=F,position="dodge")+facet_wrap(~Genus,scales="free")+
  theme_bw()+
  theme(panel.border=element_blank(),
        panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        axis.line=element_line(colour="black"),
        axis.text.x=element_text(size=16),
        axis.text.y=element_text(size=16))+
  theme(axis.title.x=element_text(size=18),axis.title.y=element_text(size=18))+
  theme(legend.text=element_text(size=16,face="italic"),legend.title=element_text(size=16))+
  theme(plot.title=element_text(lineheight=1,face="bold",size=19))+
  theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        axis.line = element_line(colour = "black",size=0.9),
        axis.text.x = element_blank(),
        axis.text.y = element_text( hjust = 1,size=8),
        axis.title=element_text(size=12,face="bold"),
        strip.text = element_text(size=8,face="bold"))+
  scale_colour_manual(values=c("darkred","darkolivegreen","dodgerblue2"))+scale_y_continuous(trans='sqrt')

```

```{r, echo= TRUE, fig.align='center'}
statisticValue<-function(value=NULL){
  if(value>0.1){
    return (paste("=",round(value,digits=2)))
  }
  else if(value>=0.05){
    return (paste("=",round(value,digits=2)))
  }
  else if(value>=0.01){
    return (paste("<0.05"))
  }
  else if(value>=0.001){
    return (paste("<0.01"))
  }
  else if(value>=0.0001){
    return(paste("<0.001"))
  }
  else{return (paste("<0.0001"))}
}
boxplotNumericByGroup<-function(mydata,category,variable,nbvariable,test,Rank=NULL){
  # if(is.null(Rank)){
  #    title<-paste(as.character(variable), " by ",category)
  #    fileForOutput<-paste(as.character(variable),"by",category,sep="_")
  #  }else{
  Phylum<-unique(as.vector(mydata[,Rank]))
  title<-paste(as.character(variable), " of\n ",Phylum,as.character(Rank)," by ",category)
  fileForOutput<-paste(as.character(variable),"of",Phylum,as.character(Rank),"by",category,sep="_")
  #  }
  numberOfLevels<-length(unique(mydata[,category]))
  colorsPlotRight<-c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7","#999999")
  require(gridExtra)
  if(numberOfLevels==2){
    test<-t.test(mydata[,variable]~mydata[,category])
    testString<-paste("Students t-test. p-value",statisticValue(test$p.value))
  }
  if(numberOfLevels>=3){
    require(MASS)
    glm.nb.model<-glm.nb(mydata[,nbvariable]~mydata[,category],method="glm.fit")
    glm.nb.aov<-aov(glm.nb.model)
    test<-aov(mydata[,variable]~mydata[,category])
    tukey.test<-TukeyHSD(test)
    print(tukey.test$mydata)
    mymatrix<-tukey.test$mydata
    mymatrix<-mymatrix[,c("diff","p adj")]
    mymatrix<-as.matrix(mymatrix)
    #mymatrix<-round(mymatrix,3)
    for(i in 1:nrow(mymatrix)){
      print(mymatrix[i,"p adj"])
      mymatrix[i,"p adj"]<-statisticValue(as.numeric(mymatrix[i,"p adj"]))
    }
    text2.df<-as.table(mymatrix)
    testString<-paste("ANOVA PR(>F)", statisticValue(summary(test)[[1]][["Pr(>F)"]][[1]]))
    testString<-paste(testString,"\n","NegBin ANOVA PR(>F)",statisticValue(summary(glm.nb.aov)[[1]][["Pr(>F)"]][[1]]))
  }
  #  if(test=="lm"){
  #    model<-lm(mydata[,variable]~mydata[,category])
  #  }
  mydata$xaxis<-mydata[,category]
  mydata$yvalue<-mydata[,variable]
  p<-ggplot(mydata,aes(x=xaxis,y=yvalue,fill=as.factor(xaxis)))+
    geom_boxplot()+geom_jitter(color="DarkRed")+
    #ggtitle(title)+
    xlab(category)+ylab(variable)+
    ylim(min(mydata$yvalue-1),1)+
    scale_fill_manual(values=colorsPlotRight[1:numberOfLevels])+
    theme(legend.position=c(1,1),legend.justification=c(1,1))+
    #annotate("text",x=numberOfLevels/2.5,y=0.5,label=testString,size=3)+
    annotate("text",x=1,y=0,label=testString,size=3)+
    annotate("text",y=0.4,x=1.5,label=title,size=4)
  plotRight<-ggplot(mydata,aes(yvalue,fill=xaxis))+geom_density(alpha=.5)+
    coord_flip()+scale_fill_manual(values=colorsPlotRight[1:numberOfLevels])+
    theme(legend.position="none")+
    xlim(min(mydata$yvalue-1),max(mydata$yvalue+1))
  if(numberOfLevels>=3){
    p<-p+annotation_custom(tableGrob(text2.df), ymin=min(mydata$yvalue)-1, ymax=min(mydata$yvalue), xmax=numberOfLevels/1.2, xmin=numberOfLevels/2)
  }  #p2<-tableGrob(text2.df)
  #grid.arrange(p2,p,main="prova",ncol=2)
  fileForPlot <- paste(fileForOutput,".pdf")
  pdf(fileForPlot,paper="a4r")
  grid.arrange(p,plotRight,nrow=1,ncol=2,widths=c(4,1),heights=c(4))
  #p2<-ggplot(p2)
  #ggsave(filename = fileForPlot,dpi=600,width=11, height=8.5)
  dev.off()
  grid.arrange(p,plotRight,nrow=1,ncol=2,widths=c(4,1),heights=c(4))
  #print(p2)
  #return(p2)
}
compareTwoGroups<-function(mydata=NULL,variable=NULL,category1=NULL,category2=NULL,fileForPlot=NULL,minCounts=500,maxAlpha=0.01, design=NULL){
  fileForPlot=paste("NegBin_DiffTest_",as.character(deparse(substitute(mydata))),"_",variable,"_",category1,"vs",category2,".pdf",sep="")
  print(paste("Output file for Plots: ",fileForPlot))
  fileForTable=paste("NegBin_DiffTest_",as.character(deparse(substitute(mydata))),"_",eval(variable),"_",eval(category1),"vs",eval(category2),".txt",sep="")
  print(paste("Output file for Table: ",fileForTable))
  stringForTitle=paste(variable,"/",category1,"vs",category2)
  kostic <- mydata
  LastRank<-rank_names(mydata)[length(rank_names(mydata))][[1]]
  require(phyloseq)
  require(DESeq2)
  kostic <- prune_samples(sample_sums(kostic) > minCounts, kostic)
  diagdds = phyloseq_to_deseq2(kostic, as.formula(design))
  #print(diagdds)
  gm_mean = function(x, na.rm=TRUE){
    exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
  }
  geoMeans = apply(counts(diagdds), 1, gm_mean)
  diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)
  #colData(ddsHTSeq)$condition<-factor(colData(ddsHTSeq)$condition, levels=c("untreated","treated"))
  #colData(diagdds)$condition<-factor(colData(diagdds)$condition,levels=c(category1,category2))
  diagdds = DESeq(diagdds, fitType="parametric",test="Wald")
  res=results(diagdds,contrast=c(variable,category1,category2))
  #res=results(diagdds)
  print(res)
  res = res[order(res$padj, na.last=NA), ]
  sigtab = res[(res$padj < maxAlpha), ]
  sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(kostic)[rownames(sigtab), ], "matrix"))
  sigtab$OtuID<-rownames(sigtab)
  head(sigtab)
  write.table(sigtab,file=fileForTable,sep="\t")
  #Cleanup for Positive enrichment in csigtabarcinoma
  posigtab=sigtab
  #posigtab = sigtab[sigtab[, "log2FoldChange"] > 0, ]
  #posigtab = posigtab[, c("baseMean", "log2FoldChange", "lfcSE", "padj", "Phylum", "Class", "Family", "Genus")]
  posigtab = posigtab[, c("baseMean", "log2FoldChange", "lfcSE", "padj", rank_names(mydata))]
  library("ggplot2")
  theme_set(theme_bw())
  scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
  }

  sigtabgen=sigtab
  
  #sigtabgen = subset(sigtab, !is.na(Genus))
  #sigtabgen = subset(sigtabgen, sigtabgen$Genus != "unclassified")
  # Phylum order
  x = tapply(sigtabgen$log2FoldChange, sigtabgen$Phylum, function(x) max(x))
  x = sort(x, TRUE)
  sigtabgen$Phylum = factor(as.character(sigtabgen$Phylum), levels=names(x))
  # Genus order
  if(as.character(LastRank)== "Genus"){
    sigtabgen$LastRank<-sigtabgen[,"Genus"]
  }else{
    sigtabgen$LastRank<-paste(sigtabgen[,"Genus"]," ",sigtabgen[,as.character(LastRank)])
  }
  x = tapply(sigtabgen$log2FoldChange, sigtabgen$LastRank, function(x) max(x))
  x = sort(x, TRUE)
  
  sigtabgen$LastRank = factor(as.character(sigtabgen$LastRank), levels=names(x))
  sigtabgen$log2Counts<-log2(sigtabgen$baseMean)
  sigtabgen$alpha<- 1 - sigtabgen$padj
  #pdf(fileForPlot)
  p<-ggplot(sigtabgen,aes(x=LastRank,y=log2FoldChange))
  p<-p+geom_point(aes(colour=Phylum,size=log2Counts,alpha=alpha))
  p<-p+scale_size_continuous(range=c(1,20))
  #+geom_point(aes(size=sigtabgen$log2Counts))+scale_size_area()
  p<-p+theme(axis.text.x=element_text(angle=-90,hjust=0,vjust=0.5,size=10))
  p<-p+theme(legend.key.size=unit(1,"cm"))
  p<-p+ ggtitle(paste(stringForTitle," Data:",as.character(deparse(substitute(mydata))))) +
    theme(plot.title = element_text(lineheight=.7, face="bold"))
  print(p)  
  #ggplot(sigtabgen, aes(x=Genus, y=log2FoldChange, color=Phylum)) + geom_point(size=6)+scale_size(range=c(1,5))+
  # theme(axis.text.x = element_text(angle = -90, hjust = 0,size=3, vjust=0.5), legend.key.size=unit(0.5,"cm"),legend.text=element_text(size=3))
  ggsave(filename = fileForPlot,dpi=600,width=11, height=8.5)
  return(sigtab)
  #dev.off()
}

### Let's move one step further and screen all genus for significant difference by using a more complex (and adequate) statistical framework using DESeq2 package and negative binomial distribution fits to detect over(under)-represented genus in a dichotomic condition.

compareTwoGroups(mydata=x.3.0.genus,variable="Site",category1="Teeth",category2="Saliva",design=~Site,maxAlpha=0.01)
compareTwoGroups(mydata=x.3.0.genus,variable="Site",category1="Teeth",category2="Cheek",design=~Site,maxAlpha=0.01)



```

In the plots we can observed plenty of differences in the genus abundance between the sampling site, comparing Teeth vs Saliva and Teeth vs Cheek. There are differences across the whole microbiome. The more increased genus is Corynebacterium in both plots, also Selenomonas,Cardiobacterium and Actinomyces are increased too.

```{r, echo= TRUE, fig.align='center'}

my.subset<-subset(ps.melt.x.3.0.genus,ps.melt.x.3.0.genus$Genus=="Cardiobacterium")
p1<-boxplotNumericByGroup(my.subset,category="Site",variable="AbundanceProportionLog",nbvariable="Abundance",Rank="Genus")

my.subset<-subset(ps.melt.x.3.0.genus,ps.melt.x.3.0.genus$Genus=="Selenomonas")
p1<-boxplotNumericByGroup(my.subset,category="Site",variable="AbundanceProportionLog",nbvariable="Abundance",Rank="Genus")

my.subset<-subset(ps.melt.x.3.0.genus,ps.melt.x.3.0.genus$Genus=="Corynebacterium")
p1<-boxplotNumericByGroup(my.subset,category="Site",variable="AbundanceProportionLog",nbvariable="Abundance",Rank="Genus")

my.subset<-subset(ps.melt.x.3.0.genus,ps.melt.x.3.0.genus$Genus=="Actinomyces")
p1<-boxplotNumericByGroup(my.subset,category="Site",variable="AbundanceProportionLog",nbvariable="Abundance",Rank="Genus")

```

### 2.Which are the bacterial genus that are enriched in teeth samples compared to both cheek and saliva?

We have performed and ANOVA test and we observed significantly differences in the for genus *(Selenomonas, Corynebacterium, Cardiobacterium* and also in *Actinomyces*) regarding the sampling site. We can affirm that this genus are enriched in teeth samples.

The bacterial genus are similar to those described in the oral microbiome in human immunodeficiency virus (HIV)-positive individuals. Where the predominant genera in plaque (constituting \>5 % mean relative abundance) were *Streptococcus, Corynebacterium, Prevotella, Capnocytophaga, Actinomyces, Veillonella and Selenomonas.*

-   *Kistler, James O et al. "The oral microbiome in human immunodeficiency virus (HIV)-positive individuals." Journal of medical microbiology vol. 64,9 (2015): 1094-1101.*

### 3.Create stacked sample barplots for each sampling site, separately, showing phylum distributions.

```{r, echo= TRUE, fig.align='center', fig.cap= 'Barplots for each sampling site (Teeth, Saliva and Cheek) showing phylum distributions.'}
my.levels=orderBy(~-AbundanceProportion,data=ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$Phylum=="Bacteroidetes",])$SampleID
ps.melt.x.3.0.phylum$SampleID<-factor(ps.melt.x.3.0.phylum$SampleID,
                                      levels=my.levels,ordered=T)

### Define Level Order for Y axis (phylum,genus,species)
my.levels<-orderBy(~-AbundanceProportion,data=ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$SampleID == as.character(my.levels[1]),])$Phylum
ps.melt.x.3.0.phylum$Phylum<-factor(ps.melt.x.3.0.phylum$Phylum,levels=my.levels,ordered=T)

### We also take care of colors
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7","#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

# We create objects for samples of each sampling site 
Teeth <- ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$Site=="Teeth",]
Saliva <- ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$Site=="Saliva",]
Cheek <- ps.melt.x.3.0.phylum[ps.melt.x.3.0.phylum$Site=="Cheek",]


## Barplot of phylum distributions in Teeth samples
colorCount=length(unique(ps.melt.x.3.0.phylum$Phylum))
getPalette=colorRampPalette(brewer.pal(12,"Set3"))
p<-ggplot(Teeth,aes(x=SampleID,y=AbundanceProportion,order=Phylum))
p+geom_bar(stat="identity",aes(fill=Phylum))+ theme(axis.text.x = element_text(size=rel(0.8),angle =70, hjust = 1))+labs(title= "Phylum Level in Teeth Samples")+ 
  scale_fill_manual(values=getPalette(colorCount))


## Barplot of phylum distributions in Saliva samples
colorCount=length(unique(ps.melt.x.3.0.phylum$Phylum))
getPalette=colorRampPalette(brewer.pal(12,"Set3"))
p<-ggplot(Saliva,aes(x=SampleID,y=AbundanceProportion,order=Phylum))
p+geom_bar(stat="identity",aes(fill=Phylum))+ theme(axis.text.x = element_text(size=rel(0.9),angle = 45, hjust = 1))+labs(title= "Phylum Level in Saliva Samples")+ 
  scale_fill_manual(values=getPalette(colorCount))

## Barplot of phylum distributions in Cheek samples
colorCount=length(unique(ps.melt.x.3.0.phylum$Phylum))
getPalette=colorRampPalette(brewer.pal(12,"Set3"))
p<-ggplot(Cheek,aes(x=SampleID,y=AbundanceProportion,order=Phylum))
p+geom_bar(stat="identity",aes(fill=Phylum))+ theme(axis.text.x = element_text(size=rel(0.8),angle = 55, hjust = 1))+labs(title= "Phylum Level in Cheek Samples")+ 
  scale_fill_manual(values=getPalette(colorCount))


```

### ORDINATION ANALYSIS

### 4.Run ordination analysis using PCoA/WUnifrac distances on the dataset and map sampling site and HIV Status as color/shape.

### ORDINATION ANALYSIS USING PCoA

```{r, echo= TRUE, fig.align='center'}
### Start Ordination Analysis
### For this we need to use ecological distance. We will be using both Bray-Curtis distance
### and WUnifrac phylogenetic distance for this tutorial

#Let put our data aside
x.4.0<-data
### calculations for this particular reason

#####Intended for de novo OTU-picking only, we keep OTUs that appear at least in two different samples
wh0=genefilter_sample(x.4.0,filterfun_sample(function(x) x>0), A=0.01*nsamples(x.4.0))
x.4.0<-prune_taxa(wh0,x.4.0)
#x.4.0<-tax_glom(x.4.0,taxrank="Genus")

### We transform the data to proportion abundances
x.4.0 = transform_sample_counts(x.4.0, function(x) ((x/sum(x))))

### We will be using NMDS which used an initial random seed for iteration
### Since we want to be able to repeat the same analysis we fix a seed for analysis
set.seed(12345)

### NMDS Ordination with Bray-Curtis distance
x.4.0.ord<-ordinate(x.4.0,"NMDS",distance="bray",trymax=200)

### Let's take a look at stressplot. Is it good?
stressplot(x.4.0.ord)
### Simple Plot
p.4.0.samples=plot_ordination(x.4.0,x.4.0.ord)
p.4.0.samples

### Let's map some metadata with coloured and ellipses
### Using HIV STatus and Site
p.4.0.samples  + geom_point(size = 2.2,aes(color=HIVStatus,fill=Site,shape=HIVStatus)) +ggtitle("Unconstrained NMDS(Bray-Curtis)")+
  stat_ellipse(geom="polygon",alpha=0.25,aes(fill=Site),level=0.95)+
  scale_colour_manual(values=c("darkred","darkolivegreen","dodgerblue2"))+
  theme_bw()+
  theme(panel.border=element_blank(),panel.grid.major=element_blank(),panel.grid.minor=element_blank(),axis.line=element_line(colour="black"),axis.text.x=element_text(size=16),axis.text.y=element_text(size=16))+
  theme(axis.title.x=element_text(size=18),axis.title.y=element_text(size=18))+
  theme(legend.text=element_text(size=16,face="italic"),legend.title=element_text(size=16))+
  theme(plot.title=element_text(lineheight=1,face="bold",size=19))
```

#### ORDINATION ANALYSIS USING WUnifrac

```{r, echo= TRUE, fig.align='center'}
### Start Ordination Analysis
### For this we need to use ecological distance. We will be using WUnifrac phylogenetic distance for this tutorial

## Phylogenetic OTU-OTU relationship
### using unifrac distances
x.4.0.ord<-ordinate(x.4.0,"NMDS",distance="wunifrac")
capture.output(file="NMDS_WUnifrac_proportions_ordinfo.txt",x.4.0.ord)
pdf("NMDS_WUnifrac_proportions_stressplot.pdf")
stressplot(x.4.0.ord)
dev.off()

### Simple Plot
p.4.0.samples=plot_ordination(x.4.0,x.4.0.ord)
p.4.0.samples
########

p.4.0.samples  + geom_point(size = 2.5,aes(color=HIVStatus,fill=Site,shape=HIVStatus,size=Age)) +ggtitle("Unconstrained NMDS(Weighted UniFrac)")+
  stat_ellipse(geom="polygon",alpha=0.25,aes(fill=Site),level=0.95)+
  scale_colour_manual(values=c("darkred","darkolivegreen","dodgerblue2"))+
  theme_bw()+
  theme(panel.border=element_blank(),panel.grid.major=element_blank(),panel.grid.minor=element_blank(),axis.line=element_line(colour="black"),axis.text.x=element_text(size=16),axis.text.y=element_text(size=16))+
  theme(axis.title.x=element_text(size=18),axis.title.y=element_text(size=18))+
  theme(legend.text=element_text(size=16,face="italic"),legend.title=element_text(size=16))+
  theme(plot.title=element_text(lineheight=1,face="bold",size=19))


```
